### PR TITLE
Add tests fresh express

### PR DESCRIPTION
**Pull request by FrankEssenberger**
## Proposed Changes

If a fresh express project is created also two tests are added:
- sample integration test: Starting the express app locally and doing a ping
- sample unit test: Just a simple assertion 

Also two jest configs are included for the unit and integration tests. The configs mainly set the reporting diretory to the S4hana_pipeline so that everything is in line with the pipeline.

## Type of Changes

- [ ] Bug Fix
- [x] New Feature
- [ ] Documentation Update

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
~~- [ ] I have added necessary documentation~~

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -48,13 +49,11 @@
     "/npm-shrinkwrap.json",
     "/oclif.manifest.json"
   ],
-  "homepage": "https://developers.sap.com/topics/cloud-sdk.html",
+  "homepage": "https://github.wdf.sap.corp/MA/sdk-js",
```
Comment (florian-richter) [inline package.json]:
```suggestion
  "homepage": "https://developers.sap.com/topics/cloud-sdk.html",
```

File: package.json [hunk 2]
```diff
@@ -1,43 +1,44 @@
 {
-  "name": "@sap-cloud-sdk/cli",
+  "name": "sap-cloud-sdk-cli",
```
Comment (florian-richter) [inline package.json]:
```suggestion
  "name": "@sap-cloud-sdk/cli",
```

File: package.json [hunk 3]
```diff
@@ -1,43 +1,44 @@
...
   "description": "CLI for the SAP Cloud SDK for JavaScript",
   "version": "0.0.1",
   "author": "SAP",
   "bin": {
     "sap-cloud-sdk": "./bin/run"
   },
-  "bugs": "https://github.com/SAP/cloud-sdk-cli/issues",
+  "bugs": "https://github.wdf.sap.corp/MA/sdk-js/issues",
```
Comment (florian-richter) [inline package.json]:
```suggestion
  "bugs": "https://github.com/SAP/cloud-sdk-cli/issues",
```

File: package.json [hunk 4]
```diff
@@ -48,13 +49,11 @@
...
-  "homepage": "https://developers.sap.com/topics/cloud-sdk.html",
+  "homepage": "https://github.wdf.sap.corp/MA/sdk-js",
   "keywords": [
-    "sap-cloud-sdk",
-    "cloud-sdk-cli",
-    "sap-cloud-platform"
+    "oclif"
   ],
-  "license": "Apache-2.0",
+  "license": "MIT",
```
Comment (florian-richter) [inline package.json]:
```suggestion
  "license": "Apache-2.0",
```

File: package.json [hunk 5]
```diff
@@ -48,13 +49,11 @@
     "/npm-shrinkwrap.json",
     "/oclif.manifest.json"
   ],
-  "homepage": "https://developers.sap.com/topics/cloud-sdk.html",
+  "homepage": "https://github.wdf.sap.corp/MA/sdk-js",
   "keywords": [
-    "sap-cloud-sdk",
-    "cloud-sdk-cli",
-    "sap-cloud-platform"
+    "oclif"
```
Comment (florian-richter) [inline package.json]:
```suggestion
    "sap-cloud-sdk",
    "cloud-sdk-cli",
    "sap-cloud-platform"
```

File: package.json [hunk 6]
```diff
@@ -72,7 +71,7 @@
       "@oclif/plugin-warn-if-update-available"
     ]
   },
-  "repository": "https://github.com/SAP/cloud-sdk-cli",
+  "repository": "https://github.wdf.sap.corp/MA/sdk-js",
```
Comment (florian-richter) [inline package.json]:
```suggestion
  "repository": "https://github.com/SAP/cloud-sdk-cli",
```

File: package-lock.json [hunk 7]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "@sap-cloud-sdk/cli",
+  "name": "sap-cloud-sdk-cli",
```
Comment (florian-richter) [inline package-lock.json]:
```suggestion
  "name": "@sap-cloud-sdk/cli",
```

File: jest.config.js [hunk 8]
```diff
@@ -1,5 +1,9 @@
 module.exports = {
   preset: 'ts-jest',
   testEnvironment: 'node',
-  modulePathIgnorePatterns: ['<rootDir>/lib', '<rootDir>/bin', '<rootDir>/tmp', '<rootDir>/test/nest', '<rootDir>/test/express']
+  modulePathIgnorePatterns: ['<rootDir>/lib', '<rootDir>/bin', '<rootDir>/tmp', '<rootDir>/test/nest', '<rootDir>/test/express'],
+  transform: {"^.+\\.tsx?$": "ts-jest"},
+  testMatch: ["**/*.spec.ts"],
+  roots: ["test"],
+  reporters: ['default']
```
Comment (florian-richter) [inline jest.config.js]: Is this change necessary? Seems like it's all the default values and can be omitted

File: src/commands/init.ts [hunk 9]
```diff
@@ -72,13 +72,15 @@ export default class Init extends Command {
...
+    const initializationType = await this.determineInitializationType(flags);
+    await this.initProject(flags, initializationType);
 
     const options = await this.getOptions();
 
     try {
       cli.action.start('Reading templates');
-      const files = readTemplates([path.resolve(__dirname, '..', 'templates', 'init')], flags.projectDir);
+      const files = readTemplates([path.resolve(__dirname, '..', 'templates', 'init')], flags.projectDir, ['test']);
+      this.readTestSampleFiles(flags.projectDir, initializationType).forEach(file => files.push(file));
```
Comment (florian-richter) [inline src/commands/init.ts]:
``` ts
const files = [
  ...readTemplates([path.resolve(__dirname, '..', 'templates', 'init')], flags.projectDir, ['test']),
  ...this.readTestSampleFiles(flags.projectDir, initializationType)
];
```

File: src/templates/init/jest.config.js [hunk 10]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  transform: {"^.+\\.tsx?$": "ts-jest"},
+  roots:["integration-tests","unit-tests"]
```
Comment (florian-richter) [inline src/templates/init/jest.config.js]: Please be careful with formatting in the templates
``` ts
transform: { "^.+\\.tsx?$": "ts-jest" },
roots: ["integration-tests", "unit-tests"]
```

File: src/templates/init/jest.integration-test.config.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  transform: {"^.+\\.tsx?$": "ts-jest"},
+  roots: ["integration-tests"],
+  reporters: [
+    'default',
+    [ "jest-junit", { suiteName: "jest tests",outputDirectory:"s4hana_pipeline/reports/backend-integration/" } ]
+  ]
```
Comment (florian-richter) [inline src/templates/init/jest.integration-test.config.js]: Formatting suggestion
``` ts
[
  "jest-junit",
  {
    suiteName: "jest tests",
    outputDirectory: "s4hana_pipeline/reports/backend-integration/" 
  }
]
```

File: src/templates/init/jest.integration-test.config.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  transform: {"^.+\\.tsx?$": "ts-jest"},
```
Comment (florian-richter) [inline src/templates/init/jest.integration-test.config.js]:
```suggestion
  transform: { "^.+\\.tsx?$": "ts-jest" },
```

File: src/templates/init/jest.unit-test.config.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  transform: {"^.+\\.tsx?$": "ts-jest"},
+  roots: ["unit-tests"],
+  reporters: [
+    'default',
+    [ "jest-junit", { suiteName: "jest tests",outputDirectory:"s4hana_pipeline/reports/backend-unit/" } ]
+  ]
```
Comment (florian-richter) [inline src/templates/init/jest.unit-test.config.js]: See above

File: src/templates/init/test/sample-integration-test.spec.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+describe('Sample integration tests', () => {
+  const https = require('http');
+  const server = require('../app.js');
+  const port = 3000;
+  let serverInstance:any;
```
Comment (florian-richter) [inline src/templates/init/test/sample-integration-test.spec.ts]: I would not use TypeScript for the tests in this project as it introduces a new dependency. Consider changing it to javascript.

File: src/templates/init/test/sample-unit-test.spec.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+describe('Sample unit test', () => {
+  it('should do some unit test',  () => {
+    expect(2+2).toBe(4);
```
Comment (florian-richter) [inline src/templates/init/test/sample-unit-test.spec.ts]: I would add an actual unit test. A good candidate would be something like `normalizePort` from `./bin/www`.

File: src/utils/initialization-helper.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    'test': 'jest'
+  },
+  devDependencies: {
+    'jest': '^24.9.0',
+    'ts-jest': '^24.0.2',
+    '@types/jest': '^24.0.18',
+    'ts-node': '^8.3.0',
+    'tslint': '^5.18.0',
+    'typescript': '^3.5.3',
+    'jest-junit': '^8.0.0'
```
Comment (florian-richter) [inline src/utils/initialization-helper.ts]: As described above, I would not use typescript here. Only devDependencies I see as MUST HAVE are `jest` and `jest-junit`. Adding `eslint` (instead of `tslint`) might be a NICE TO HAVE.

File: src/utils/templates.ts [hunk 17]
```diff
@@ -8,11 +8,11 @@ import { compile } from 'handlebars';
 import * as path from 'path';
 import { CopyDescriptor } from './copy-list';
 
-export function readTemplates(fromDirectory: string[], toDirectory: string): CopyDescriptor[] {
+export function readTemplates(fromDirectory: string[], toDirectory: string, excludedDirs: string[] = []): CopyDescriptor[] {
```
Comment (florian-richter) [inline src/utils/templates.ts]: Consider changing the function definition to or similar
``` ts
interface templateParam {
  from: string[];
  to: string;
  exclude?: string[];
}

export function readTemplates({ from, to, exclude = [] }: templateParam): CopyDescriptor[] {
```

File: test/init.spec.ts [hunk 18]
```diff
@@ -21,12 +21,15 @@ jest.mock('@oclif/command', () => {
   return command;
 });
 
+import execa = require('execa');
 import * as fs from 'fs-extra';
 import * as path from 'path';
 import Init from '../src/commands/init';
 
+const foo: string = 'bar';
```
Comment (florian-richter) [inline test/init.spec.ts]: This line should be removed, I think?

File: test/init.spec.ts [hunk 19]
```diff
@@ -54,6 +57,30 @@ describe('Init', () => {
...
+    const projectDir = path.resolve(pathPrefix, 'full-init-with-test');
+    if (fs.existsSync(projectDir)) {
+      fs.removeSync(projectDir);
+    }
+
+    const argv = ['--projectName=testingApp', '--startCommand="npm start"', '--frontendScripts', '--initWithExpress', `--projectDir=${projectDir}`];
+    await Init.run(argv);
+
+    // execute the local tests
+    await execa('npm', ['i'], {cwd: projectDir});
```
Comment (florian-richter) [inline test/init.spec.ts]:
```suggestion
    await execa('npm', ['install'], { cwd: projectDir });
```

File: test/init.spec.ts [hunk 20]
```diff
@@ -54,6 +57,30 @@ describe('Init', () => {
...
+    // execute the ci scripts and check if the reports are written
+    await execa('npm', ['run', 'ci-integration-test'], { cwd: projectDir });
+    const pathBackendIntegration = path.resolve(projectDir, 's4hana_pipeline', 'reports', 'backend-integration');
+    expect(fs.readdirSync(pathBackendIntegration).length).not.toBe(0);
+
+    await execa('npm', ['run', 'ci-backend-unit-test'], { cwd: projectDir });
+    const pathBackendUnit = path.resolve(projectDir, 's4hana_pipeline', 'reports', 'backend-unit');
+    expect(fs.readdirSync(pathBackendUnit).length).not.toBe(0);
+
+  }, 120000);
```
Comment (florian-richter) [inline test/init.spec.ts]: This can possibly reduced to `60000` again.

File: jest.config.js (hunk 8 above)
Comment (FrankEssenberger) [reply jest.config.js]: they are almost not necessary, only the root so that the sample test files in templates are not executed.

File: src/templates/init/jest.integration-test.config.js (hunk 12 above)
Comment (FrankEssenberger) [reply src/templates/init/jest.integration-test.config.js]: after removing TS no transformer needed anymore.

File: src/templates/init/jest.integration-test.config.js (hunk 11 above)
Comment (FrankEssenberger) [reply src/templates/init/jest.integration-test.config.js]: hm, The suiteName is optinal and I removved it. It would prefe a single line

File: src/templates/init/jest.config.js [hunk 21]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  transform: { '^.+\\.tsx?$': 'ts-jest' },
```
Comment (florian-richter) [inline src/templates/init/jest.config.js]: `ts-jest` is no longer used.

File: src/templates/init/test/sample-unit-test.spec.js [hunk 22]
```diff
@@ -0,0 +1,5 @@
+describe('Sample unit test', () => {
+  it('should do some unit test', () => {
```
Comment (florian-richter) [inline src/templates/init/test/sample-unit-test.spec.js]:
```suggestion
  it('should add two numbers', () => {
```

File: src/templates/init/test/test-file-info.ts [hunk 23]
```diff
@@ -0,0 +1,15 @@
+interface testFileInfo {
+  fileName: string;
+  targetFolder: string;
+}
+
+export const testFiles: testFileInfo[] = [
```
Comment (florian-richter) [inline src/templates/init/test/test-file-info.ts]: This should
- not be part of the `templates/` folder (actually it should not be necessary at all)
- be implicitly clear by the folder structure (I'm pretty sure `fs` and `path` have the functions you need)

File: src/commands/init.ts [hunk 24]
```diff
@@ -3,19 +3,19 @@
...
 
-const backendScripts = {
+const backendBuildScripts = {
   'ci-build': 'echo "Use this to compile or minify your application"',
-  'ci-package': 'echo "Copy all deployment-relevant files to the `deployment` folder"',
-  'ci-backend-unit-test':
-    'echo "Test your application and write results in a JUnit format to `s4hana_pipeline/reports/backend-unit/` and coverage in a cobertura format to `s4hana_pipeline/reports/coverage/backend-unit/`"',
-  'ci-integration-test':
-    'echo "Test your application and write results in a JUnit format to `s4hana_pipeline/reports/backend-integration/` and coverage in a cobertura format to `s4hana_pipeline/reports/coverage/backend-integration/`"'
+  'ci-package': 'echo "Copy all deployment-relevant files to the `deployment` folder"'
```
Comment (florian-richter) [inline src/commands/init.ts]: Maybe all scripts should be moved to the initHelper.ts

File: src/templates/init/jest.config.js (hunk 21 above)
Comment (FrankEssenberger) [reply src/templates/init/jest.config.js]: yes

File: src/commands/init.ts (hunk 24 above)
Comment (FrankEssenberger) [reply src/commands/init.ts]: move it too the helper

File: src/templates/init/test/test-file-info.ts (hunk 23 above)
Comment (FrankEssenberger) [reply src/templates/init/test/test-file-info.ts]: yes, used the basic reader method and model the strucutre right away

File: src/commands/init.ts [hunk 25]
```diff
@@ -72,13 +58,19 @@ export default class Init extends Command {
...
     const options = await this.getOptions();
 
     try {
       cli.action.start('Reading templates');
-      const files = readTemplates([path.resolve(__dirname, '..', 'templates', 'init')], flags.projectDir);
+      const excludes = initializationType === InitializationType.existingProject ? ['test'] : [];
+      const files = readTemplates({
+        from: [path.resolve(__dirname, '..', 'templates', 'init')],
+        to: flags.projectDir,
+        exclude: excludes
```
Comment (florian-richter) [inline src/commands/init.ts]: Very minor thing but I would shorten `initializationType` to `initType` (same for the uppercase one) and move the `initType === InitType.existingProject ? ['test'] : [];` to line 72 so

``` ts
exclude: initType === InitType.existingProject ? ['test'] : []
```

File: src/commands/init.ts [hunk 26]
```diff
@@ -103,31 +95,43 @@ export default class Init extends Command {
     }
   }
 
-  private async initExpressProject() {
-    const { flags } = this.parse(Init);
+  private async determineInitializationType(flags: OutputFlags<typeof Init.flags>): Promise<InitializationType> {
```
Comment (florian-richter) [inline src/commands/init.ts]: I assume there are testing reasons why you pass flags as param instead of putting `const { flags } = this.parse(Init);` into the function?

I like passing the flags generally, but the type def does look nice. If we keep it as the parameters we should be consistent between all the methods and also put something like `type Flags = OutputFlags<typeof Init.flags>` somewhere to shorten the param def.
